server/src/main/java/org/elasticsearch/common/metrics/ExponentialBucketHistogram.java

```diff
@@ -0,0 +1,102 @@
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.common.metrics;
+
+import java.util.Arrays;
+import java.util.concurrent.atomic.LongAdder;
+
+/**
+ * A histogram with a fixed number of buckets of exponentially increasing width.
+ * <p>
+ * The bucket boundaries are defined by increasing powers of two, e.g.
+ * <code>
+ *     (-&infin;, 1), [1, 2), [2, 4), [4, 8), ..., [2^({@link #BUCKET_COUNT}-2), &infin;)
+ * </code>
+ * There are {@link #BUCKET_COUNT} buckets.
+ */
+public class ExponentialBucketHistogram {
+
+    public static int[] getBucketUpperBounds() {
+        int[] bounds = new int[17];
+        for (int i = 0; i < bounds.length; i++) {
+            bounds[i] = 1 << i;
+        }
+        return bounds;
+    }
+
+    private static int getBucket(long observedValue) {
+        if (observedValue <= 0) {
+            return 0;
+        } else if (LAST_BUCKET_LOWER_BOUND <= observedValue) {
+            return BUCKET_COUNT - 1;
+        } else {
+            return Long.SIZE - Long.numberOfLeadingZeros(observedValue);
+        }
+    }
+
+    public static final int BUCKET_COUNT = getBucketUpperBounds().length + 1;
+
+    private static final long LAST_BUCKET_LOWER_BOUND = getBucketUpperBounds()[BUCKET_COUNT - 2];
+
+    private final LongAdder[] buckets;
+
+    public ExponentialBucketHistogram() {
+        buckets = new LongAdder[BUCKET_COUNT];
+        for (int i = 0; i < BUCKET_COUNT; i++) {
+            buckets[i] = new LongAdder();
+        }
+    }
+
+    public void addObservation(long observedValue) {
+        buckets[getBucket(observedValue)].increment();
+    }
+
+    /**
+     * @return An array of frequencies of handling times in buckets with upper bounds as returned by {@link #getBucketUpperBounds()}, plus
+     *         an extra bucket for handling times longer than the longest upper bound.
+     */
+    public long[] getHistogram() {
+        final long[] histogram = new long[BUCKET_COUNT];
+        for (int i = 0; i < BUCKET_COUNT; i++) {
+            histogram[i] = buckets[i].longValue();
+        }
+        return histogram;
+    }
+
+    /**
+     * Calculate the Nth percentile value
+     *
+     * @param percentile The percentile as a fraction (in [0, 1.0])
+     * @return A value greater than the specified fraction of values in the histogram
+     */
+    public long getPercentile(float percentile) {
+        assert percentile >= 0 && percentile <= 1;
+        final long[] snapshot = getHistogram();
+        final long totalCount = Arrays.stream(snapshot).reduce(0L, Long::sum);
+        long percentileIndex = (long) Math.ceil(totalCount * percentile);
+        for (int i = 0; i < BUCKET_COUNT; i++) {
+            percentileIndex -= snapshot[i];
+            if (percentileIndex <= 0) {
+                return getBucketUpperBounds()[i];
```

Comment:
I think this means we'll get 2x steps in the values reported. I.e., the 90th percentile will be 2048 or 4096 but never anything in between?

I guess this speaks to reporting just the current value instead as a starting point?

Reply:
Yes it is a downside to the exponential bucket histogram. We could perhaps use a different exponent, or linear buckets?

I thought the value in the histogram over the "current" value is that , because we only get a measure every 60s, the "current" value might not tell the whole story (e.g. with current or p99 we might see a momentary spike in latency e.g. due to GC or something but with e.g. p90 we give that a chance to be diluted by an otherwise healthy rate of queue processing.